packages/console/src/console.js

```diff
@@ -0,0 +1,359 @@
+
+export const consoleOmittedProperties = freeze([
+  'memory', // Chrome
+  'exception', // FF, MDN
+  '_ignoreErrors', // Node
+  '_stderr', // Node
+  '_stderrErrorHandler', // Node
+  '_stdout', // Node
+  '_stdoutErrorHandler', // Node
+  '_times', // Node
+  'context', // Chrome, Node
+  'record', // Safari
+  'recordEnd', // Safari
+  'screenshot', // Safari
+  // A variety of symbols also seen on Node
+]);
+
+/**
+ * @typedef {readonly [string, ...any[]]} LogRecord
+ *
+ * @typedef {Object} LoggingConsoleKit
+ * @property {Console} loggingConsole
+ * @property {() => readonly LogRecord[]} takeLog
+ */
+
+/**
+ * A mock console that just accumulates the contents of all whitelisted calls,
+ * making them available to callers of `takeLog()`. Calling `takeLog()`
+ * consumes these, so later calls to `takeLog()` will only provide a log of
+ * calls that have happened since then.
+ *
+ * A logging console also implements the custom `rememberErrorInfo` method,
+ * which the `@agoric/assert` module feature tests for, so it captures and
+ * reports error annotations as well. Unlike the causalConsole below, the
+ * logging console does not delay reporting these annotations. It just
+ * immediately adds to the log what it was asked to remember.
+ *
+ * @returns {LoggingConsoleKit}
+ */
+const makeLoggingConsoleKit = () => {
+  // Not frozen!
+  let logArray = [];
+
+  const loggingConsole = fromEntries(
+    consoleWhitelist.map(name => {
+      // Use an arrow function so that it doesn't come with its own name in
+      // its printed form. Instead, we're hoping that tooling uses only
+      // the `.name` property set below.
+      const method = (...args) => {
+        logArray.push([name, ...args]);
+      };
+      defineProperty(method, 'name', { value: name });
+      return [name, freeze(method)];
+    }),
+  );
+  /** @type {RememberErrorInfo} */
+  const rememberErrorInfo = (error, errorInfoKind, getLogArgs) => {
+    const logArgs = getLogArgs();
+    logArray.push([error, errorInfoKind, ...logArgs]);
+  };
+  loggingConsole.rememberErrorInfo = rememberErrorInfo;
+  freeze(loggingConsole);
+
+  const takeLog = () => {
+    const result = freeze(logArray);
+    logArray = [];
+    return result;
+  };
+  freeze(takeLog);
+
+  return freeze({ loggingConsole, takeLog });
+};
+freeze(makeLoggingConsoleKit);
+export { makeLoggingConsoleKit };
+
+const defaultGetStackString = error => {
+  if (!('stack' in error)) {
+    return '';
+  }
+  const stackString = `${error.stack}`;
+  const pos = stackString.indexOf('\n');
+  if (stackString.startsWith(' ') || pos === -1) {
+    return stackString;
+  }
+  return stackString.slice(pos + 1); // exclude the initial newline
+};
+
+/**
+ * Makes a causal console wrapper of a base console, where the causal wrappper
+ * uses `decodeConsole` to recognize
+ *
+ * @param {Console} baseConsole
+ * @returns {Console}
+ */
+const makeCausalConsole = (
+  baseConsole,
+  getStackString = defaultGetStackString,
```

Comment:
Perhaps a good spot for a forward-looking options bag here.

Reply:
Done